kms/capi/capi.go

```diff
@@ -0,0 +1,754 @@
+		storeName = "My"
+	}
+
+	st, err := windows.CertOpenStore(
+		certStoreProvSystem,
+		0,
+		0,
+		certStoreLocation,
+		uintptr(unsafe.Pointer(wide(storeName))))
+	if err != nil {
+		return nil, fmt.Errorf("CertOpenStore for the %v store %v returned: %w", storeLocation, storeName, err)
+	}
+
+	var certHandle *windows.CertContext
+
+	switch {
+	case sha1Hash != "":
+		sha1Hash = strings.TrimPrefix(sha1Hash, "0x") // Support specifying the hash as 0x like with serial
+
+		sha1Bytes, err := hex.DecodeString(sha1Hash)
+		if err != nil {
+			return nil, fmt.Errorf("%s must be in hex format: %w", HashArg, err)
+		}
+		searchData := CERT_ID_KEYIDORHASH{
+			idChoice: CERT_ID_SHA1_HASH,
+			KeyIDOrHash: CRYPTOAPI_BLOB{
+				len:  uint32(len(sha1Bytes)),
+				data: uintptr(unsafe.Pointer(&sha1Bytes[0])),
+			},
+		}
+		certHandle, err = findCertificateInStore(st,
+			encodingX509ASN|encodingPKCS7,
+			0,
+			findCertID,
+			uintptr(unsafe.Pointer(&searchData)), nil)
+		if err != nil {
+			return nil, fmt.Errorf("findCertificateInStore failed: %w", err)
+		}
+		if certHandle == nil {
+			return nil, fmt.Errorf("certificate with %v=%s not found", HashArg, keyID)
+		}
+		defer windows.CertFreeCertificateContext(certHandle)
+		return certContextToX509(certHandle)
+	case keyID != "":
+		keyID = strings.TrimPrefix(keyID, "0x") // Support specifying the hash as 0x like with serial
+
+		keyIDBytes, err := hex.DecodeString(keyID)
+		if err != nil {
+			return nil, fmt.Errorf("%v must be in hex format: %w", KeyIDArg, err)
+		}
+		searchData := CERT_ID_KEYIDORHASH{
+			idChoice: CERT_ID_KEY_IDENTIFIER,
+			KeyIDOrHash: CRYPTOAPI_BLOB{
+				len:  uint32(len(keyIDBytes)),
+				data: uintptr(unsafe.Pointer(&keyIDBytes[0])),
+			},
+		}
+		certHandle, err = findCertificateInStore(st,
+			encodingX509ASN|encodingPKCS7,
+			0,
+			findCertID,
+			uintptr(unsafe.Pointer(&searchData)), nil)
+		if err != nil {
+			return nil, fmt.Errorf("findCertificateInStore failed: %w", err)
+		}
+		if certHandle == nil {
+			return nil, fmt.Errorf("certificate with %v=%s not found", KeyIDArg, keyID)
+		}
+		defer windows.CertFreeCertificateContext(certHandle)
+		return certContextToX509(certHandle)
+	case issuerName != "" && serialNumber != "":
+		var serialBytes []byte
+		if strings.HasPrefix(serialNumber, "0x") {
+			serialNumber = strings.TrimPrefix(serialNumber, "0x")
+			serialNumber = strings.TrimPrefix(serialNumber, "00") // Comparison fails if leading 00 is not removed
+			serialBytes, err = hex.DecodeString(serialNumber)
+			if err != nil {
+				return nil, fmt.Errorf("invalid hex format for %v: %w", SerialNumberArg, err)
+			}
+		} else {
+			bi := new(big.Int)
+			bi, ok := bi.SetString(serialNumber, 10)
+			if !ok {
+				return nil, fmt.Errorf("invalid %v - must be in hex or integer format", SerialNumberArg)
+			}
+			serialBytes = bi.Bytes()
+		}
+		var prevCert *windows.CertContext
+		for {
+			certHandle, err = findCertificateInStore(st,
```

Comment:
In the previous cases, you are freeing the handle with `windows.CertFreeCertificateContext(certHandle)`, shouldn't we free them here too? Note that you're iterating using them.

Reply:
The handle is freed below on L615

As far as im aware `findCertificateInStore` frees `prevCert` before returning a new one, so we only need to free the handle if `findCertificateInStore` returns us something and we DONT continue iterating through (ie we find our cert).

If `findCertificateInStore` exhausts its search it frees the handle and returns nil